File: ios/NitroGoogleCast/Converters/MediaQueueContainerMetadata+toGckMediaQueueContainerMetadata.swift
import Foundation
import GoogleCast
import NitroModules

/// Converts a generated `MediaQueueContainerMetadata` struct into a
/// `GCKMediaQueueContainerMetadata` via `GCKMediaQueueContainerMetadataBuilder`.
///
/// Reverse lives in `GCKMediaQueueContainerMetadata+toMediaQueueContainerMetadata.swift`.
/// `containerType` is non-optional in GCK; absent maps to the builder default `Generic`, which
/// the reverse maps to `generic` (so absent `containerType` normalizes to `generic`).
/// `authors`/`narrators`/`publisher`/`releaseDate` are not modelled by our struct.
extension MediaQueueContainerMetadata {
  func toGckMediaQueueContainerMetadata() -> GCKMediaQueueContainerMetadata {
    let builder = GCKMediaQueueContainerMetadataBuilder(
      containerType: containerType?.toGckContainerType() ?? .generic)
    if let title { builder.title = title }
    if let containerDuration { builder.containerDuration = containerDuration }
    if let containerImages { builder.containerImages = containerImages.map { $0.toGckImage() } }
    if let sections { builder.sections = sections.map { $0.toGckMediaMetadata() } }
    return builder.build()
  }
}

File: android/src/main/java/com/margelo/nitro/googlecast/converters/MediaQueueContainerMetadata+toGckMediaQueueContainerMetadata.kt
package com.margelo.nitro.googlecast.converters

import com.margelo.nitro.googlecast.MediaQueueContainerMetadata
import com.margelo.nitro.googlecast.MediaQueueContainerType
import com.google.android.gms.cast.MediaQueueContainerMetadata as GckMediaQueueContainerMetadata

/**
 * Converts a generated [MediaQueueContainerMetadata] struct into a Google Cast
 * [GckMediaQueueContainerMetadata].
 *
 * The reverse lives in `GckMediaQueueContainerMetadata+toMediaQueueContainerMetadata.kt`.
 * `containerType` is mapped BY VALUE. `containerDuration` is seconds (`double`) on both sides.
 */
internal fun MediaQueueContainerMetadata.toGckMediaQueueContainerMetadata(): GckMediaQueueContainerMetadata {
  val builder = GckMediaQueueContainerMetadata.Builder()
  containerType?.let {
    val gckType = when (it) {
      MediaQueueContainerType.GENERIC -> GckMediaQueueContainerMetadata.MEDIA_QUEUE_CONTAINER_TYPE_GENERIC
      MediaQueueContainerType.AUDIOBOOK -> GckMediaQueueContainerMetadata.MEDIA_QUEUE_CONTAINER_TYPE_AUDIO_BOOK
    }
    builder.setContainerType(gckType)
  }
  title?.let { builder.setTitle(it) }
  containerDuration?.let { builder.setContainerDuration(it) }
  containerImages?.let { images -> builder.setContainerImages(images.map { it.toGckWebImage() }) }
  sections?.let { metadatas -> builder.setSections(metadatas.map { it.toGckMediaMetadata() }) }
  return builder.build()
}
